File: packages/@alicloud/ros-cdk-core/lib/helpers-internal/string-specializer.ts
import * as cxapi from '@alicloud/ros-cdk-cxapi';
import { RosPseudo } from '../ros-pseudo';
import { Stack } from '../stack';
import { Token } from '../token';

/**
 * A "replace-all" function that doesn't require us escaping a literal string to a regex
 */
function replaceAll(str: string, search: string, replace: string) {
  return str.split(search).join(replace);
}

export function generateRandomString(length: number): string {
  const characters = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let result = '';
  const charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
}

export class StringSpecializer {
  /**
   * Validate that the given string does not contain tokens
   */
  public static validateNoTokens(str: string, what: string) {
    if (Token.isUnresolved(str)) {
      throw new Error(`${what} may not contain tokens; only the following literal placeholder strings are allowed: ` + [
        '${Qualifier}',
        cxapi.EnvironmentPlaceholders.CURRENT_REGION,
        cxapi.EnvironmentPlaceholders.CURRENT_ACCOUNT,
      ].join(', ') + `. Got: ${str}`);
    }
  }

  constructor(private readonly stack: Stack, private readonly qualifier: string) { }

  /**
   * Function to replace placeholders in the input string as much as possible
   *
   * We replace:
   * - ${Qualifier}: always
   * - ${ALIYUN::AccountId}, ${ALIYUN::Region}: only if we have the actual values available
   */
  public specialize(str: string): string {
    str = replaceAll(str, '${Qualifier}', this.qualifier);
    return cxapi.EnvironmentPlaceholders.replace(str, {
      region: resolvedOr(this.stack.region, cxapi.EnvironmentPlaceholders.CURRENT_REGION),
      accountId: resolvedOr(this.stack.account, cxapi.EnvironmentPlaceholders.CURRENT_ACCOUNT),
    });
  }

  /**
   * Specialize the given string, make sure it doesn't contain tokens
   */
  public specializeNoTokens(str: string, what: string): string {
    StringSpecializer.validateNoTokens(str, what);
    return this.specialize(str);
  }

  /**
   * Specialize only the qualifier
   */
  public qualifierOnly(str: string): string {
    return replaceAll(str, '${Qualifier}', this.qualifier);
  }
}

/**
 * Return the given value if resolved or fall back to a default
 */
export function resolvedOr<A>(x: string, def: A): string | A {
  return Token.isUnresolved(x) ? def : x;
}

const ASSET_TOKENS = ['${ALIYUN::Region}', '${ALIYUN::AccountId}'];
const ROS_TOKENS = [RosPseudo.region, RosPseudo.accountId];

/**
 * Replaces ROS Tokens with corresponding Asset Tokens.
 */
export function translateRosTokenToAssetToken(arn: string) {
  for (let i = 0; i < ROS_TOKENS.length; i++) {
    arn = replaceAll(arn, ROS_TOKENS[i], ASSET_TOKENS[i]);
  }
  return arn;
}

/**
 * Replaces Asset Tokens with corresponding ROS Tokens.
 */
export function translateAssetTokenToRosToken(arn: string) {
  for (let i = 0; i < ASSET_TOKENS.length; i++) {
    arn = replaceAll(arn, ASSET_TOKENS[i], ROS_TOKENS[i]);
  }
  return arn;
}

File: multiple-languages/java/ros-cdk-core/src/main/java/com/aliyun/ros/cdk/core/StringSpecializer.java
package com.aliyun.ros.cdk.core;

@javax.annotation.Generated(value = "jsii-pacmak/1.85.0 (build 08ee592)", date = "2024-06-06T03:35:12.153Z")
@software.amazon.jsii.Jsii(module = com.aliyun.ros.cdk.core.$Module.class, fqn = "@alicloud/ros-cdk-core.StringSpecializer")
public class StringSpecializer extends software.amazon.jsii.JsiiObject {

    protected StringSpecializer(final software.amazon.jsii.JsiiObjectRef objRef) {
        super(objRef);
    }

    protected StringSpecializer(final software.amazon.jsii.JsiiObject.InitializationMode initializationMode) {
        super(initializationMode);
    }

    public StringSpecializer(final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.core.Stack stack, final @org.jetbrains.annotations.NotNull java.lang.String qualifier) {
        super(software.amazon.jsii.JsiiObject.InitializationMode.JSII);
        software.amazon.jsii.JsiiEngine.getInstance().createNewObject(this, new Object[] { java.util.Objects.requireNonNull(stack, "stack is required"), java.util.Objects.requireNonNull(qualifier, "qualifier is required") });
    }

    /**
     * Validate that the given string does not contain tokens.
     * <p>
     * @param s This parameter is required.
     * @param what This parameter is required.
     */
    public static void validateNoTokens(final @org.jetbrains.annotations.NotNull java.lang.String s, final @org.jetbrains.annotations.NotNull java.lang.String what) {
        software.amazon.jsii.JsiiObject.jsiiStaticCall(com.aliyun.ros.cdk.core.StringSpecializer.class, "validateNoTokens", software.amazon.jsii.NativeType.VOID, new Object[] { java.util.Objects.requireNonNull(s, "s is required"), java.util.Objects.requireNonNull(what, "what is required") });
    }

    /**
     * Specialize only the qualifier.
     * <p>
     * @param s This parameter is required.
     */
    public @org.jetbrains.annotations.NotNull java.lang.String qualifierOnly(final @org.jetbrains.annotations.NotNull java.lang.String s) {
        return software.amazon.jsii.Kernel.call(this, "qualifierOnly", software.amazon.jsii.NativeType.forClass(java.lang.String.class), new Object[] { java.util.Objects.requireNonNull(s, "s is required") });
    }

    /**
     * Function to replace placeholders in the input string as much as possible.
     * <p>
     * We replace:
     * <p>
     * <ul>
     * <li>${Qualifier}: always</li>
     * <li>${ALIYUN::AccountId}, ${ALIYUN::Region}: only if we have the actual values available</li>
     * </ul>
     * <p>
     * @param s This parameter is required.
     */
    public @org.jetbrains.annotations.NotNull java.lang.String specialize(final @org.jetbrains.annotations.NotNull java.lang.String s) {
        return software.amazon.jsii.Kernel.call(this, "specialize", software.amazon.jsii.NativeType.forClass(java.lang.String.class), new Object[] { java.util.Objects.requireNonNull(s, "s is required") });
    }

    /**
     * Specialize the given string, make sure it doesn't contain tokens.
     * <p>
     * @param s This parameter is required.
     * @param what This parameter is required.
     */
    public @org.jetbrains.annotations.NotNull java.lang.String specializeNoTokens(final @org.jetbrains.annotations.NotNull java.lang.String s, final @org.jetbrains.annotations.NotNull java.lang.String what) {
        return software.amazon.jsii.Kernel.call(this, "specializeNoTokens", software.amazon.jsii.NativeType.forClass(java.lang.String.class), new Object[] { java.util.Objects.requireNonNull(s, "s is required"), java.util.Objects.requireNonNull(what, "what is required") });
    }
}
